### PR TITLE
Implement IntoFuture type inference

### PR DESCRIPTION
**Pull request by jridgewell**
One of my projects is using [IntoFuture](https://doc.rust-lang.org/std/future/trait.IntoFuture.html) to make our async code a little less verbose. However, rust-analyzer can't infer the output type of an await expression if the value uses `IntoFuture` to convert into another type. So we're getting `{unknown}` types everywhere since switching.

`foo.await` itself [desugars](https://github.com/rust-lang/rust/blob/e4417cf020fbcd6182c11637bc6b8694434bd81a/compiler/rustc_ast_lowering/src/expr.rs#L644-L658) into a `match into_future(foo) {}`, with every `Future` impl getting a [default](https://github.com/rust-lang/rust/blob/e4417cf020fbcd6182c11637bc6b8694434bd81a/library/core/src/future/into_future.rs#L131-L139) `IntoFuture` implementation. I'm not sure if we want to disable the old `future_trait` paths, since this only recently [stabilize](https://github.com/rust-lang/rust/pull/98718).

### CODE REVIEW

File: crates/hir/src/lib.rs [hunk 1]
```diff
@@ -2780,6 +2780,8 @@ impl Type {
     /// Checks that particular type `ty` implements `std::future::Future`.
     /// This function is used in `.await` syntax completion.
     pub fn impls_future(&self, db: &dyn HirDatabase) -> bool {
+        // FIXME: This should be checking for IntoFuture trait, but I don't know how to find the
+        // right TraitId in this crate.
```
Comment (Veykril) [inline crates/hir/src/lib.rs:2784]: `IntoFuture::into_future` is tagged as the `into_future` lang item, so you could look for that function, then turn it into an AssocItem and fetch its trait container which should be the IntoFuture trait

Comment (jridgewell) [reply crates/hir/src/lib.rs]: Done (and could use a review, I'm very new to this). Any suggestion for how to fix the `resolve_await_to_poll` problem?

File: crates/hir/src/lib.rs [hunk 2]
```diff
@@ -2780,17 +2780,28 @@ impl Type {
...
+
+        let trait_ = match trait_ {
             Some(it) => it,
             None => return false,
         };
 
         let canonical_ty =
             Canonical { value: self.ty.clone(), binders: CanonicalVarKinds::empty(Interner) };
-        method_resolution::implements_trait(&canonical_ty, db, self.env.clone(), std_future_trait)
+        method_resolution::implements_trait(&canonical_ty, db, self.env.clone(), trait_)
```
Comment (flodiebold) [inline crates/hir/src/lib.rs:2804]: should this function be renamed `impls_into_future`? Certainly the doc string needs to be adapted.

File: crates/hir/src/source_analyzer.rs [hunk 3]
```diff
@@ -269,6 +269,8 @@ impl SourceAnalyzer {
         db: &dyn HirDatabase,
         await_expr: &ast::AwaitExpr,
     ) -> Option<FunctionId> {
+        // FIXME This should be pointing to the poll of IntoFuture::Output's Future impl, but I
+        // don't know how to resolve the Output type so that we can query for its poll method.
```
[the remark this answers is not in the document]
Comment (flodiebold) [reply crates/hir/src/source_analyzer.rs:273]: we actually have a `normalize_trait_assoc_type` method on `Type` that you should be able to use for this

[the remark this answers is not in the document]
Comment (jridgewell) [reply crates/hir/src/source_analyzer.rs]: Done.

**Comment by Veykril** [comment]
Thanks!
@bors r+

**Pull request merged**